xain_fl/coordinator/coordinator.py

```diff
@@ -457,30 +469,31 @@ def _handle_end_training_round(
         # The round is over. Run the aggregation
         if self.round.is_finished():
             logger.info(
-                "Running aggregation for round", current_round=self.current_round
+                "Running aggregation for round", current_round=current_round
             )
 
             multiple_model_weights: List[ndarray]
             aggregation_data: List[int]
             multiple_model_weights, aggregation_data = self.round.get_weight_updates()
-            self.weights = self.aggregator.aggregate(
+            weights = self.aggregator.aggregate(
                 multiple_model_weights=multiple_model_weights,
                 aggregation_data=aggregation_data,
             )
+            self.session.set_weights(weights)
             self.global_weights_writer.write_weights(
-                self.current_round + 1, self.weights
+                current_round + 1, weights
             )
 
             # update the round or finish the training session
-            if self.current_round >= self.num_rounds - 1:
-                logger.info("Last round over", round=self.current_round)
-                self.state = State.FINISHED
-                self._write_metrics_fail_silently("coordinator", {"state": self.state})
+            if current_round >= self.num_rounds - 1:
+                logger.info("Last round over", round=current_round)
+                self.session.set_state(State.FINISHED)
+                self._write_metrics_fail_silently("coordinator", {"state": State.FINISHED})
             else:
-                self.current_round += 1
-                self.epoch_base += self.epochs_current_round
+                self.session.next_round()
```

Comment:
This kind of update thing is still racy, since the lock is released between `session.next_round()` and `session.add_epochs()`. It this something we're ok with?

Reply:
More generally actually, the session state can be updated from multiple threads in the `_handle_end_training_round` method, so I wonder if we should not put the session itself behind a lock

Reply:
yes I had hoped it would be possible to not have a giant lock in the coordinator (or have it encapsulated away) but maybe it's unavoidable after all!